### ISSUE TITLE
[API Proposal]: Allow IReadOnlyDictionary GetValueOrDefault parameter key to be nulllable

### ISSUE THREAD

**Issue by schehlmj** [dotnet/runtime#67738]
### Background and motivation

Dictionaries do not allow keys to be null. Because of that, it can be assured that calling GetValueOrDefault with a null key would not return a value, therefore, it is safe that if a null value for key is passed in, that returning the default is acceptable.

This code change allows code like:
var nullableInt = (int?)null;
var dict = new Dictionary<int, string>();
var foo = nullableInt.HasValue ? dict.TryGetValueOrDefault(nullableInt.Value) : null;

to be like:
var nullableInt = (int?)null;
var dict = new Dictionary<int, string>();
var foo = dict.TryGetValueOrDefault(nullableInt);

### API Proposal

```C#
using System.Collections.ObjectModel;
using System.Diagnostics.CodeAnalysis;

namespace System.Collections.Generic
{
    public static class CollectionExtensions
    {
        public static TValue? GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary!!, TKey key)
        {
            if (key == null) return default!;
            return dictionary.GetValueOrDefault(key, default!);
        }

        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary!!, TKey key, TValue defaultValue)
        {
            if (key == null) return defaultValue;
            TValue? value;
            return dictionary.TryGetValue(key, out value) ? value : defaultValue;
        }

        public static TValue? GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary!!, TKey? key) where TKey : struct
        {
            if (!key.HasValue) return default!;
            return dictionary.GetValueOrDefault(key, default!);
        }
        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary!!, TKey? key, TValue defaultValue) where TKey : struct
        {
            if (!key.HasValue) return defaultValue;
            TValue? value;
            return dictionary.TryGetValue(key.Value, out value) ? value : defaultValue;
        }
  ....
}
```

### API Usage

```C#
var nullableInt = (int?)null;
var dict = new Dictionary<int, string>();
var foo = dict.TryGetValueOrDefault(nullableInt);  // no ternary operator is ended to handle when nullableInt is null
```

### Risks

ArgumentNullException would no longer be thrown by TryGetValueOrDefault if a null key for an object was passed in.  It is possible that if existing code was expecting the ArgumentNullException  to be thrown when a null was passed in would break, however, the general expectation of TryGetValueOrDefault is that the "Try" should not allow exceptions but should return Default.

**Comment by eiriktsarpalis** [comment]
If I'm honest I don't believe there is a lot of value in adding this extensions, since they can be trivially expressed as one-liner expressions or factor into user-defined extension methods. What might be more useful is supporting dictionaries with null keys (technically this is possible using IEqualityComparer implementations that support null values, but most dictionary implementations hardcode a null check regardless of this)

**Comment by schehlmj** [comment]
I agree that a dictionary supporting null keys would be be better, but even with that, I believe allowing a nullable key to be passed in when the Dictionary key is a struct (and so can't be null) is still needed.

The proposed would provide a lot of value in that they would be expected behavior (principle of least surprise).

I guess, in my line of thinking, I would say, because the "GetValueOrDefault" method says "...OrDefault"  I expect the method to return default rather than an exception.  And so the value is following the principle of least surprise.

**Comment by eiriktsarpalis** [comment]
> I believe allowing a nullable key to be passed in when the Dictionary key is a struct (and so can't be null) is still needed.

Why would you want to do that when `TKey` itself is not nullable? If it's null then clearly it can't be part of the dictionary (even if the implementation hypothetically supported null keys) so this is simply validating the key parameter before passing it in. I wouldn't support the inclusion of such an extension method in BCL.

> I guess, in my line of thinking, I would say, because the "GetValueOrDefault" method says "...OrDefault" I expect the method to return default rather than an exception. And so the value is following the principle of least surprise.

Doesn't `TryGetValue` help in that regard? I normally just write `TryGetValue(key, out var value) ? value : default;` in such cases.

**Comment by schehlmj** [comment]
Hi @eiriktsarpalis , 

In my cases, I have a Dictionary to find objects based on their database primary key.  Naturally, the primary key is non-nullable, and in my cases its an int.  When finding values from that dictionary, I very often have an int? because foreign keys are often nullable.  

Logically, it is very clear that a Dictionary<int, object> cannot contain a key that is null, and therefore, any tries to search for a null key in the Dictionary should return the default value as the name of the "GetValueOrDefault" implies.

The .Net languages, especially C#, have been moving toward greater convenience, flexibility and ease of use-- and this seems to fall in line perfectly with that.  

- There is already precedent for Nullables to be handled specially for convenience:
  - eg. instead of having to write:  var foo = nullableIntValue.HasValue ? intValue == nullableIntValue.Value ? false;
  -  you can write:  var foo = intValue == nullableIntValue;
- Having GetValueOrDefault handle nulls for structs would follow that same convenience for developers:
   -  current: var foo = nullableInt.HasValue ? dict.**GetValueOrDefault**(nullableInt.Value) : null;
   - proposed: var foo =  dict.**GetValueOrDefault**(nullableInt);   
- The code to implement this change in the extension method is basically 1 line of code
- Existing user code cannot be broken by this change, except if contains a user-defined extension with same name/sig

Hopefully, the above helps sway you to supporting this convenience, if not, we can close this issue.

On a separate, related note, should I create a separate [API Proposal] for your possible suggestion that Dictionary<TKey, TValue> should allow null keys and not throw ArgumentNullException?

**Comment by eiriktsarpalis** [comment]
> Logically, it is very clear that a Dictionary<int, object> cannot contain a key that is null, and therefore, any tries to search for a null key in the Dictionary should return the default value as the name of the "GetValueOrDefault" implies.

Fundamentally the proposed methods are meant to handle input sanitization responsibilities, in a very particular way. I don't think using extension methods over dictionary is the right way to achieve this (wrapping or extending your dictionary using a custom class might be preferable). One might argue that we need a "GetValueOrDefault" that accepts an `object` key just so that _every_ conceivable value can be handled, but ultimately extension methods need to have parity with the contract offered by the type they are meant to extend. The contract of a `Dictionary<TKey, TValue>` is that it can only hold keys of type `TKey : notnull` and nothing else.

> On a separate, related note, should I create a separate [API Proposal] for your possible suggestion that Dictionary<TKey, TValue> should allow null keys and not throw ArgumentNullException?

It might be possible to consider a proposal for null key support via custom `IEqualityComparer<TKey>` instances, but it won't be something that we'd be able to support by default.

With that said, I'm going to close this issue, I would not support adding the proposed extension methods.
